dist/doc/spec.adoc

```diff
@@ -0,0 +1,1270 @@
+  greater than 0 or the end of the array if the given step is less than 0.
+* If a negative stop position is given, it is calculated as the total length
+  of the array plus the given stop position.
+* If no stop position is given, it is assumed to be the length of the array if
+  the given step is greater than 0 or 0 if the given step is less than 0.
+* If the given step is omitted, it it assumed to be 1.
+* If the given step is 0, an `invalid-value` error MUST be raised.
+* If the element being sliced is not an array, the result is `null`.
+* If the element being sliced is an array and yields no results, the result
+  MUST be an empty array.
+
+=== Examples
+
+[source%unbreakable]
+----
+  search(foo["bar"], {"foo": {"bar": 21}}) -> 21
+  search([0:4:1], [0, 1, 2, 3]) -> [0, 1, 2, 3]
+  search([0:4], [0, 1, 2, 3]) -> [0, 1, 2, 3]
+  search([0:3], [0, 1, 2, 3]) -> [0, 1, 2]
+  search([:2], [0, 1, 2, 3]) -> [0, 1]
+  search([::2], [0, 1, 2, 3]) -> [0, 2]
+  search([::-1], [0, 1, 2, 3]) -> [3, 2, 1, 0]
+  search([-2:], [0, 1, 2, 3]) -> [2, 3]
+----
+
+=== Flatten Operator
+
+When the character sequence `[]` is provided as a bracket specifier, then
+a flattening operation occurs on the current result.  The flattening operator
+will merge sub-lists in the current result into a single list.  The flattening
+operator has the following semantics:
+
+* Create an empty result list.
+* Iterate over the elements of the current result.
+* If the current element is not a list, add to the end of the result list.
+* If the current element is a list, add each element of the current element
+  to the end of the result list.
+* The result list is now the new current result.
+
+Once the flattening operation has been performed, subsequent operations
+are projected onto the flattened list with the same semantics as a
+wildcard expression.  Thus the difference between `[*]` and `[]` is that
+`[]` will first flatten sub-lists in the current result.
+
+=== Examples
+
+[source%unbreakable]
+----
+  search([0], ["first", "second", "third"]) -> "first"
+  search([-1], ["first", "second", "third"]) -> "third"
+  search([100], ["first", "second", "third"]) -> null
+  search(foo[0], {"foo": ["first", "second", "third"]}) -> "first"
+  search(foo[100], {"foo": ["first", "second", "third"]}) -> null
+  search(foo[0][0], {"foo": [[0, 1], [1, 2]]}) -> 0
+  search(foo[], {"foo": [[0, 1], [1, 2]]}) -> [0,1,1,2]
+----
+
+== Operators
+
+=== Comparison Operators
+
+The following comparison operators are supported:
+
+* '=' / `==`, test for equality.
+* `!=`, "<>", test for inequality.
+* `<`, less than.
+* `\<=`, less than or equal to.
+* `>`, greater than.
+* `>=`, greater than or equal to.
+
+* If both operands are numbers, a numeric comparison is performed.
+* If both operands are strings, they are compared as strings, based on the values of the Unicode code points they contain.
+* If operands are mixed types, follow the <<Type Coercion,type coercion>> rules.
```

Comment:
I agree with what you have mentioned but that needs to be implemented.

Reply:
Hmm.  good points.  The spec says that "Any other type shall be coerced to a number".  But it also says that converting arrays and objects to numbers is not supported.  
I think that ordered comparison of arrays and objects is meaningless and so we should do whatever is simplest.  
Either:

- Make the ordered comparison of arrays and objects an error
OR
- convert a populated array and populated object to one
- convert an empty array and empty object to zero
OR
- convert to string using toString() and compare strings

I prefer the 1/0 approach. It would be somewhat consistent with the boolean logic where empty objects/arrays are false and populated ones are true.  

What do you think?